### ISSUE TITLE
Zcash crashes with segfault or hangs in uninterruptible sleep when attempting to send unconfirmed utxos from z-address

### ISSUE THREAD

**Issue by runn1ng** [zcash/zcash#2480]
### Describe the issue

I want to send zcash, using unconfirmed utxos.

#### If so, please list the steps to reproduce below:
1. Send zcash to fresh z-address from t-address, using ./zcash-cli z_sendmany
2. Now *before this transaction confirms*, try to send to another z-address using z_sendmany and minconf set to 0, from the address from 1
3. Zcashd crashes with segfault immediately

Sending from t-address to t-address works; sending confirmed utxo works. It has to be from z-address, and it has to be unconfirmed. (Not sure what happens when sending from z-address to t-address.)

### Expected behaviour
Zcash is sent

### Actual behaviour + errors
Zcash almost immediately terminates and writes `Segmentation fault (core dumped)`. There is no other error.

### The version of Zcash you were using:
`Zcash Daemon version v1.0.9`, from latest release tag

I built the binary myself

### Machine specs:
- OS name + version: Lubuntu
- CPU: x86_64
- RAM: 4GB (the transaction seems to be created fine, no issue with low memory normally)
- Disk size: > 100 GB free space
- Disk Type (HD/SDD): HD
- Linux kernel version (uname -a): 4.10.0
- Compiler version (gcc -version): gcc (Ubuntu 5.4.0-6ubuntu1~16.04.4) 5.4.0 20160609

### Any extra information that might be useful in the debugging process.
Last line of `~/.zcash/debug.log` is the successful transaction from before.

### Do you have a back up of `~/.zcash` directory and/or take a VM snapshot?
Not right now. If the problem is not reproducible anywhere else, I might try that.

**Comment by daira** [comment]
@runn1ng Can you upgrade to 1.0.10-1 and confirm that this error still occurs, please?

**Comment by tromer** [comment]
If the consumed note was created by an unconfirmed transaction, then no valid anchor exists yet for the JoinSplit in the new transaction. So, is the crash due to dereferencing a NULL anchor pointer?

**Comment by arielgabizon** [comment]
@runn1ng  I wonder if you could indeed upgrade to 1.0.10-1 (by sudo apt-get upgrade zcash)
to see if you still get the segfault, cause I have tried to t-addr and to z-addr from a z-addr with unconfirmed balance, and I don't get a segfault , I get no error, and then when I do zcash-cli -regtest z_getoperationresult  I get a "insufficient funds" message .
Maybe it behaves different on regtest I'll need to try on testnet

**Comment by arielgabizon** [comment]
OK @runn1ng we produced the problem on mainnet. We didn't run it on debug mode so it just got stuck - can you perhaps send us the line in the code where the segfault happened (just in case you have it)?
 The solution seems to be not to allow minconf=0 for transactions from zaddreses (indeed if you don't add the 0 in the end, it just says insufficient funds and doesn't crash)
cause as @tromer said, transactions from z-addrs should fail with 0 confirmations, since they haven't been incorporated in the note commitment tree yet, and that is necessary for the zero-knowledge proof to work.

**Comment by bitcartel** [comment]
@daira @arielgabizon So you are seeing different behaviour on regtest than mainnet?  What does "it just got stuck" mean?  Also it sounds like you weren't able to reproduce the segfault?

**Comment by arielgabizon** [comment]
Now it wasn't about regtest or mainnet, it was that it needs to get the "0" parameter for minconf in the end to get stuck. I tried now on regtest and same thing happens. 
Stuck means zcashd is not responding for example if you write `zcash-cli stop` there is no response
or if you write `zcash-cli st` and press tab it get's stuck and doesn't give you an autocomplete
I'm guessing @runn1ng ran it in some sort of debug mode and that's why he got a segfault message .
To clarify the problem happens when sending from a z-addr (not t-addr) that has notes with 0 confirmations and adding "0" in the end of z_sendmany

**Comment by arielgabizon** [comment]
Ohh small correction in regtest it simply terminates zcashd (without a message)
Btw this even happens if the z_addr had notes that were with >0 confirmations sufficing to cover the z_sendmany, but also had a note with 0 confirmations covering it

**Comment by nathan-at-least** [comment]
Fixing this segfault is important because the bug may be "deep".

However, there is a "higher" UX / interface bug here: Requesting a transfer from a z-addr's note which has 0 block confirmations is never valid. Users don't explicitly refer to notes, and therefore cannot make such a request (explicitly). So, how can we fix the UX problem (which is logically distinct from the segfault)?

Given that `MINCONF` is a _minimum_, one option is to clip the minimum to 1 effectively whenever the sending address is a z-addr. I kind of dislike this, because it's different behavior than for t-addr send addresses.

Another option, which I like more (because it addresses more cases) is to gracefully fail when valid notes can not be found. This is hard to do well, because users need to know what the problem is and how to react to it, and there could be multiple causes of note selection failure. This is a general enough solution that it deserves its own separate ticket, #2492, leaving this ticket to focus on fixing the segfault.

**Comment by daira** [comment]
"Stuck" here means that zcashd is in D state (uninterruptible sleep). That shouldn't happen at all for long periods, so it may be that the process is doing something that is undefined behaviour and that causes either a D-state sleep or a segfault depending on platform (kernel version maybe)?

**Comment by daira** [comment]
As for the UI/API problem, there are for options I can see:

1. Error if minconf=0 for z-addrs, but allow it for t-addrs.
2. Silently set minconf to 1 if given as 0 for z-addrs.
3. Error if minconf=0 for z_sendmany regardless of address type.
4. Error if minconf=0 always, even for old Bitcoin APIs

Option 4 removes useful functionality. Option 3 is inconsistent between z_sendmany and Bitcoin APIs for t-addrs. I think we should exclude these two options.

Although I'm not normally a fan of changing any parameter silently, note that option 2 does have some merit precisely because minconf is a *minimum* number of confirmations. The contract of z_sendmany says that it uses inputs with at least the given number of confirmations. An input with 1 confirmation is also an input with at least 0 confirmations, so the contract is met. Note that an input may gain a confirmation (or in rare cases, more than one) as we perform the z_sendmany anyway.

The difference between options 1 and 2 is only that option 1 treats minconf=0  as an error on the part of the caller, forcing them to make their intent clear. Although this would normally be my preference for an API, I think that in this case I've talked myself into preferring option 2 by the argument above, also because this makes the t-addr and z-addr preconditions more consistent.

Note that options 1 and 2 don't differ in behaviour in the success case, and there is no situation where a call will succeed with option 1 but not with option 2.

**Comment by daira** [comment]
I don't understand how @nathan-at-least's proposal to "gracefully fail when valid notes can not be found" actually differs from option 2.

**Comment by runn1ng** [comment]
I can confirm I get segfault when using 1.0.10-1. I use only mainnet, not regnet or testnet.

I self-built zcashd using `./zcutil/build.sh -j4`; I did not add any debugging.

I am not sure how to debug the segfault - and I am not even sure what "core dumped" means and where is it actually dumped :laughing:; if you guide me, I can help you.

I can try to use the version from APT now.

**Comment by runn1ng** [comment]
When I am using the version from APT, I also get segfault. So it behaves exactly the same for me.

**Comment by runn1ng** [comment]
About the API change - I just want to note other calls, like `z_gettotalbalance` and probably others, allow to set `minconf=0`. Not sure how zcash should deal with this.

**Comment by daira** [comment]
Most Linux distributions disable core dumps by default. To enable them, do this in the same shell before running zcashd:
```
ulimit -c unlimited
```
Since the core file will contain private keys, it would probably be better for you to extract a traceback from it yourself rather than sending the core file to us. For this it's better to use the zcashd you built yourself (the one installed from apt has no symbol information, so the traceback from that one would not be very useful).

The core file is typically called ``core`` and written in the current directory. To get a traceback from it, do:
```
gdb core
tb
```
(if I'm remembering correctly), and paste the output here.

**Comment by bitcartel** [comment]
Confirmed and fixed in PR #2526.  Source of segfault located.  Dereferencing null ptr with `->nHeight`.
```
Thread 6 "zcashd" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7f761321e700 (LWP 16640)]
0x000055d536589391 in AsyncRPCOperation_sendmany::main_impl (
    this=this@entry=0x7f75cc003430)
    at wallet/asyncrpcoperation_sendmany.cpp:592
592	                wtxHeight = mapBlockIndex[wtx.hashBlock]->nHeight;
```

**Comment by runn1ng** [comment]
Thanks, I was busy and didn't have time for locating the bug :)

**Comment by runn1ng** [comment]
:+1:

**Comment by ioptio** [comment]
hey @runn1ng has this issue been resolved for you since the 1.0.11 release?
